### PR TITLE
Refactor Units Tests

### PR DESCRIPTION
**Pull request by wavefunction91**
General refactor of unit tests to close out a number of open issues

**Notable changes:**
-  Removes `Catch::Approx` from unit tests in favor of more modern Matcher scheme
-  Adds custom matcher to improve output quality
-  Implemented a randomized polynomial scheme for the unit testing of primitive quadratures
  - Changes Gauss-Chebyshev T3 to generate its "natural" bounds of [0,1] to enable simpler derivation of appropriate testing types.
 

**Related Issues:**
- [x] Closes https://github.com/wavefunction91/IntegratorXX/issues/44
- [x] Closes https://github.com/wavefunction91/IntegratorXX/issues/62
- [x] Closes https://github.com/wavefunction91/IntegratorXX/issues/64

### CODE REVIEW

**Comment by wavefunction91** [comment]
@susilehtola I'll propagate the improved printers to #52 once this is merged in

File: include/integratorxx/quadratures/gausschebyshev3.hpp [hunk 1]
```diff
@@ -74,10 +74,12 @@ struct quadrature_traits<GaussChebyshev3<PointType, WeightType>> {
...
       wi *= std::sqrt((1.0 - xi) / xi);
 
-      // Finally, convert from [0,1] to [-1,1]
-      // TODO: This should be done externally
-      points[idx] = 2 * xi - 1.0;
-      weights[idx] = 2.0 * wi;
+      //// Finally, convert from [0,1] to [-1,1]
+      //// TODO: This should be done externally
+      //points[idx] = 2 * xi - 1.0;
+      //weights[idx] = 2.0 * wi;
```
Comment (susilehtola) [inline include/integratorxx/quadratures/gausschebyshev3.hpp:80]: Shouldn't this be removed altogether?

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -18,23 +18,27 @@ else()
   message( STATUS "Found Catch2 VERSION ${Catch2_VERSION} DIR ${Catch2_DIR}" )
 endif()
 
+find_package( Boost REQUIRED )
```
Comment (susilehtola) [inline test/CMakeLists.txt:21]: I think you should list the needed components here

File: test/composite_quadratures.cxx [hunk 3]
```diff
@@ -23,14 +26,8 @@ TEST_CASE( "Spherical Quadratures", "[sph-quad]" ) {
...
-    double res = 0;
-    for( auto i = 0; i < npts; ++i ) {
-      const auto p = s.points()[i];
-      const auto rsq = p[0]*p[0] + p[1]*p[1] + p[2]*p[2];
-      res += s.weights()[i] * std::exp(-rsq);
-    }
-
-    CHECK( res == Catch::Approx( M_PI * std::sqrt(M_PI) ) );
+    const std::string msg = "MK(150) x LL(770) Spherical";
+    test_quadrature<RadialGaussian>(msg, s, std::pow(M_PI,1.5), 1e-10);
```
Comment (susilehtola) [inline test/composite_quadratures.cxx:30]: This looks sketchy

File: test/test_functions.hpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    return eval * eval;
+  }
+};
+
+struct RadialGaussian {
+  static double evaluate(double r) {
+    return r*r * std::exp(-r*r);
+  }
+  static double evaluate(double x, double y, double z) {
+    const auto r = std::hypot(x,y,z);
```
Comment (susilehtola) [inline test/test_functions.hpp:60]: Wow there's a function for this? You learn something new every day..

File: test/composite_quadratures.cxx (hunk 3 above)
Comment (wavefunction91) [reply test/composite_quadratures.cxx:30]: How so?

File: test/test_functions.hpp (hunk 4 above)
Comment (wavefunction91) [reply test/test_functions.hpp:60]: Somewhere buried deep in the ISO C++ standard, there is a STL function to wash your car (templated over make and model, of course).

File: test/CMakeLists.txt (hunk 2 above)
Comment (wavefunction91) [reply test/CMakeLists.txt:21]: This should actually be removed. Thanks

File: include/integratorxx/quadratures/gausschebyshev3.hpp (hunk 1 above)
Comment (wavefunction91) [reply include/integratorxx/quadratures/gausschebyshev3.hpp:80]: Yes

File: test/composite_quadratures.cxx (hunk 3 above)
Comment (wavefunction91) [reply test/composite_quadratures.cxx:30]: Ah, its *wrong*... b/c $r^2$ is already in the weights, fixing

Comment (susilehtola) [reply test/composite_quadratures.cxx:30]: Well the reason I was confused was that you're integrating a radial function... but I guess the test might make sense.

Are you also testing the integrals with atomic orbitals with nonzero angular momentum?

Comment (wavefunction91) [reply test/composite_quadratures.cxx]: No, but that should be a part of another PR

File: test/1d_quadratures.cxx [hunk 5]
```diff
@@ -88,218 +52,158 @@ constexpr T ref_gaussian_int( T a, T b ) {
...
+      cp[i] = c[i] / (p-i);
+    }
+    return Polynomial::evaluate(cp, 1.0) - Polynomial::evaluate(cp, -1.0);
+  };
 
-      auto f = [=]( double x ){ return gaussian(x); };
+  // Test Quadrature for Correctness
+  using quad_type = IntegratorXX::GaussLobatto<double,double>;
+  test_random_polynomial<quad_type, Polynomial>("Gauss-Lobatto", 10, 14, 
+    [](int o){ return 2*o-1; }, // Max order 2N-3
```
Comment (susilehtola) [inline test/1d_quadratures.cxx:100]: `2o-1` vs `2N-3`?

File: test/1d_quadratures.cxx [hunk 6]
```diff
@@ -88,218 +52,158 @@ constexpr T ref_gaussian_int( T a, T b ) {
...
+  };
 
-      double res = 0.;
-      for( auto i = 0; i < quad.npts(); ++i ) {
-        res += wgt[i] * f(pts[i]);
-      }
+  // Test Quadrature for Correctness
+  using quad_type = IntegratorXX::GaussLegendre<double,double>;
+  test_random_polynomial<quad_type, Polynomial>("Gauss-Legendre", 10, 14, 
+    [](int o){ return 2*o+1; }, // Max order 2N-1
```
Comment (susilehtola) [inline test/1d_quadratures.cxx:80]: `2o+1` vs `2N-1`?

File: test/1d_quadratures.cxx [hunk 7]
```diff
@@ -88,218 +52,158 @@ constexpr T ref_gaussian_int( T a, T b ) {
...
+    return ref;
+  };
 
-    // Check that nodes are in increasing value
-    CHECK(std::is_sorted(pts.begin(), pts.end()));
+  // Test Quadrature for Correctness
+  using quad_type = IntegratorXX::GaussChebyshev1<double,double>;
+  using func_type = WeightedPolynomial<ChebyshevT1WeightFunction>;
+  test_random_polynomial<quad_type, func_type>("Gauss-Chebyshev (T1)", 10, 100, 
+    [](int o){ return 2*o+1; }, // Max order 2N-1
```
Comment (susilehtola) [inline test/1d_quadratures.cxx:127]: same here

**Review by susilehtola — APPROVED** [review]

File: test/1d_quadratures.cxx (hunk 6 above)
Comment (wavefunction91) [reply test/1d_quadratures.cxx:80]: The `p` parameter generates a polynomial of order `p-1`, and the `p`-loop is not left-inclusive, hence the "off-by-2" here and elsewhere.

File: test/1d_quadratures.cxx (hunk 5 above)
Comment (wavefunction91) [reply test/1d_quadratures.cxx:100]: See https://github.com/wavefunction91/IntegratorXX/pull/65#discussion_r1285266280

File: test/1d_quadratures.cxx (hunk 7 above)
Comment (wavefunction91) [reply test/1d_quadratures.cxx:127]: See https://github.com/wavefunction91/IntegratorXX/pull/65#discussion_r1285266280

**Pull request merged**